fitlins/workflows/base.py

```diff
@@ -150,160 +76,134 @@ def join_dict(base_dict, dict_list):
+            BIDSDataSink(base_directory=out_dir,
+                         path_patterns=contrast_plot_pattern),
+            run_without_submitting=True,
+            name='ds_{}_contrast_plots'.format(step))
+
+        if step == 'run':
+            # Select preprocessed BOLD series to analyze
+            getter = pe.Node(
+                BIDSSelect(
+                    bids_dir=bids_dir, derivatives=derivatives,
+                    selectors={
+                        'type': 'preproc', 'suffix': 'bold', 'space': space}),
+                name='getter')
+
+            # Run first level model
+            model = pe.MapNode(
+                FirstLevelModel(),
+                iterfield=['session_info', 'contrast_info', 'bold_file', 'mask_file'],
+                name='{}_model'.format(step))
+            models.append(model)
+
+            def join_dict(base_dict, dict_list):
+                return [{**base_dict, **iter_dict} for iter_dict in dict_list]
+
+            # Accumulate metadata
+            l1_metadata = pe.MapNode(
+                niu.Function(function=join_dict),
+                iterfield=['base_dict', 'dict_list'],
+                name='l1_metadata')
+
+            plot_design = pe.MapNode(
+                DesignPlot(image_type='svg'),
+                iterfield='data',
+                name='plot_design')
+
+            # Images
+            ds_design = pe.MapNode(
+                BIDSDataSink(base_directory=out_dir, fixed_entities={'type': 'design'},
+                             path_patterns=image_pattern),
+                iterfield=['entities', 'in_file'],
+                run_without_submitting=True,
+                name='ds_design')
+
+            wf.connect([
+                (loader, select_entities, [('entities', 'inlist')]),
+                (loader, model, [('session_info', 'session_info')]),
+                (select_entities, getter,  [('out', 'entities')]),
+                (getter, model, [('bold_files', 'bold_file'),
+                                 ('mask_files', 'mask_file')]),
+                (getter, l1_metadata, [('entities', 'base_dict')]),
+                (model, l1_metadata, [('contrast_metadata', 'dict_list')]),
+                (model, plot_design, [('design_matrix', 'data')]),
+                (l1_metadata, collate, [('out', 'contrast_metadata')]),
+                (select_entities, ds_design, [('out', 'entities')]),
+                (plot_design, ds_design, [('figure', 'in_file')]),
+                (select_entities, ds_model_warnings,  [('out', 'entities')]),
+            ])
+
+        #  Set up higher levels
+        else:
+            model = pe.MapNode(
+                SecondLevelModel(),
+                iterfield=['contrast_info'],
+                name='{}_model'.format(step))
+
+            wf.connect([
+                (models[-1], model, [('contrast_maps', 'stat_files')]),
+                (l1_metadata, model, [('out', 'stat_metadata')]),
```

Comment:
This seems like it can't be right for L3+.

Reply:
correct. this is something i haven't figured out yet. only works for 2 levels as of now.